### ISSUE TITLE
Round-trippable representation of source code

### ISSUE THREAD

**Issue by davidanthoff** [JuliaLang/JuliaParser.jl#22]
This might serve as a discussion place for a fully round-trippable representation of julia source code. Such a representation could potentially be useful for IDEs, source code formatting tools etc. My suggestions below are heavily inspired by the [C# Roslyn design](https://github.com/dotnet/roslyn/wiki/Roslyn%20Overview).

The first step would be to change the lexer to not ever skip any content, but return a stream of everything present in the file. Here is one possible design: 1) let the lexer return elements of a new type `SyntaxToken` that would essentially replace the current `Token` typealias and 2) represent all whitespaces, comments, new lines etc as `SyntaxTrivia` elements that are children of a `SyntaxToken`.

The `next_token` function would then return elements of type `SyntaxToken`. The stream of elements this method would return would be _exactly_ the same as is currently, so I believe the existing downstream parser implementation would not have to change much, other than being able to deal with `SyntaxToken` instead of the current Union `Token` return type. The type might look like this:

``` julia
immutable SyntaxToken
  # This should probably be a fixed size array/tuple or something
  LeadingTrivia::Array{SyntaxTrivia,1} 
  TrailingTrivia::Array{SyntaxTrivia,1}

  # Or maybe an enum? This could be values like :IdentifierToken, 
  # :FunctionKeyword, :StringLiteral etc.
  Kind::Symbol

  # This could potentially hold exactly the same thing that is currently returned by
  # next token
  Value

  # The type SyntaxSpan would just have two fields Start and End (or Length) and
  # would be the character range in the source file that is covered by this token
  Span::SyntaxSpan
end
```

The type `SyntaxTrivia` would hold things like whitespaces, comments, line endings etc.:

``` julia
immutable SyntaxTrivia
  # Might be :EndOfLine, :Whitespace, :SingleLineComment
  Kind::Symbol

  # This could be left empty for Kind of :EndOfLine or :Whitespace, but
  # might contain e.g. the actual text of a comment.
  Value

  # Parent
  Token::SyntaxToken

  Span::SyntaxSpan
end
```

The tricky thing would be that the `next_token` method would need to allocate all the syntax trivia to a `SyntaxToken`. I think in Roslyn they have some simple rules, that all syntax trivia is owned by the preceding `SyntaxToken`, with some exceptions: the first `SyntaxToken` owns all preceeding trivia, all trivia after a line break is owned by the next `SyntaxToken` and the last `SyntaxToken` owns all remaining trivia, even stuff after a line break.

But I think that from a downstream point of view, that the current parser should just be able to consume such a new token stream and ignore the trivia easily.

So this is on the lexer side. I think from there on it would make most sense to have a parallel type to the existing `Expr` type for now. In Roslyn they have a type hierarchy of an abstract class `SyntaxNode`, and then concrete types for different types of syntax. Not sure whether that is better than the route that `Expr` took with a single type. But the key thing is that each `SyntaxNode` then has a vector of all the `SyntaxNode`s that constitute it, so they retain that data.

I generally have studied the `SyntaxNode` system in Roslyn less than the lexer part, but that seems the broad design. The key feature is that one can take a parsed tree of `SyntaxNode`s, and because all the syntax trivia is also preserved, write it out to a file again and have a full fidelity round trip thing. Implementing something like a syntax formatter on top of this would obviously be much easier than the current situation.

In terms of larger architecture, I guess my proposal would be to modify the current lexer in `JuliaParser` so that it returns `SyntaxToken`s, modify the current parser so that it operates based on `SyntaxToken`s, and then potentially introduce a second parser (or an option to the existing one) to produce a tree of `SyntaxNode`, in addition to the ability to produce a tree of `Expr` objects.

I guess one key question is whether such a feature set would actually help people who are trying to implement things like IDE support... The Roslyn project powers all of Visual Studio's language support, which is massive, from syntax highlighting, to tool tips, refactoring, source formatting etc.

Oh, and sorry for not just submitting a PR for this kind of thing. But I'm a bit out of my depth when it comes to lexers/parsers... The Roslyn design just looked really good to me, so I thought I should propose something like that here and see whether it might stimulate a useful discussion and maybe the interest of someone else to implement it.

**Comment by davidanthoff** [comment]
Oh, and for anyone with a PC: you should try to install Visual Studio 2015 (I think there are free versions), then install the ".NET Compiler Platform SDK" extension and use the View->Syntax Visualizer window to look at some source tree, it is really nice and makes the whole design of Roslyn a lot easier to grasp than a lot of text.

**Comment by jakebolewski** [comment]
Wow, this is a great issue thanks for a well thought out and detailed write up.  I was thinking of a design not so heavy weight as what you proposed.  It would look like clang's source info's which give a compact representation of the token delimiters while reducing the amount of memory required to save such detailed positional info in the AST.  Clang's design requires you to re-parse if you want the token info to be actionable but for the use cases a user would be interested in (IDE support, REPL completions, error reporting) doesn't seem to be a big deal. 

I think the main blocker in developing this functionality would be outside contributors (none seem to want to modify the parser to make it better for IDE support).  I don't use an IDE so unless someone can write an amazing Vimscript plugin my willingness to work on this is low as it doesn't improve my day to day life.   I know there is an effort to make Atom a "Julia IDE", but frankly Atom's Vim emulation mode sucks.

**Comment by EricForgy** [comment]
For what it's worth, I also like this idea. I have 20+ years working with Matlab so I am addicted to having some kind of basic IDE with break points, etc., so am watching the Atom stuff with interest. To my ignorant eye, this issue looks like it would mesh nicely with those efforts.

**Comment by davidanthoff** [comment]
@one-more-minute I would be interested in your take on this proposal here. Unless you would find it useful it would probably not be worth pursuing at all. It also is probably at best a medium to long term project...

**Comment by one-more-minute** [comment]
This is definitely an interesting proposal. It's really hard to say what Julia's tooling will look like in the longer term at such an early stage, so take this with a pinch of salt, but I can try to get some initial thoughts down.

In it's current form it seems like this mainly enables source-to-source transformations – formatters and refactoring tools being the main use-cases. For that use case alone I think this proposal would be nice to have.

Another big part of the Roslyn design seems to be pulling information out of the compiler. That's probably less useful for refactoring in Julia than it would be for a static language like C#, since in general you just can't know types etc. without running the program. It could be useful for other things, like mapping `code_typed` information back onto the source for debugging. On the other hand, things like macros make that kind of thing really hard. I'm not sure it's tenable to make macro writers responsible for preserving metadata about code, comments etc., as it goes through. Getting the compiler to emit that metadata in the first place is going to be a large project as well – and for stuff like code_typed you can get 90% of the way there without it.

Given that we can live code with running programs in our IDE, I think a much more promising approach is to go the route of dynamic, as opposed to static, analysis. (As a trivial example: if you want to know the type of the variable `x`, you don't need to type-infer the whole codebase: just query the runtime with `typeof(x)`). That's much more powerful and flexible than the very limited amount of static analysis that we could do with Julia. Not being able to do static analysis is limiting in some ways, but I suspect that dynamic analysis opens up some interesting new possibilities as well.

To summarise: I think richer parsing would be really useful for refactoring tools and such, but C# and Julia being so fundamentally different means that something Roslyn-like is probably the wrong end goal. If I were reworking the parser myself I'd want to do it as part of a larger concerted effort towards refactoring tools for Julia, without worrying too much about other features that Roslyn provides.

Anyway, this is just another take from someone who isn't exactly a compiler expert himself, so like I said – pinches of salt.

**Comment by mdcfrancis** [comment]
I'd like to see this sort of parse tree transformation, having round trip AST nodes is a big step to allowing code reformatting (which doesn't use regex) and potentially refactoring of code. Having worked on a language which supported this, it was a very powerful feature. Not least that we were able to enforce a standard look and feel across our code base at 'zero' cost to the developer. 

In addition it allowed us to support in-line docs and annotations in comments - for example pointers to issues ( not sure if that is the direction we want to go for Julia, but was an interesting use case )

**Comment by davidanthoff** [comment]
I agree with @one-more-minute, that all the fancy stuff past the syntax stage of Roslyn would probably not be a match for Julia. But just for reformatting stuff at the syntax level it might still be useful. I guess another valuable perspective would be whether this might be useful for something like lint, @tonyhffong?

Another question is how formatting tools etc. would interact with this. My understanding is that all the juno stuff e.g. interacts with any julia code viz zmq. I don't know, but that might easily be too slow to have things like syntax coloring etc. depend on julia code. But one probably could also embed julia via the C interace in something like Atom, and then performance would be a different story?

**Comment by mdcfrancis** [comment]
Having used the 'lint' which was written for the language I was referencing, it was a heavy user of the AST and lint was the driving engine for the coding style. 

I agree with @one-more-minute that Julia is a different beast - I suspect if the compiler output CommentNodes and MacroNodes in the ast that would get me 99% of what I would need. 

Refactoring becomes more important in larger enterprise style projects - for those we would have to have very rich static analysis blended with some deep introspection of the type system.

One area I could see this being helpful would be for removing deprecated functions. Being able to semantically transform the parsed data to remove all the @compat ... macros while retaining the formatting would be nice.

**Comment by tonyhffong** [comment]
+1 to that. Lint can benefit hugely from a round-trippable representation.

**Comment by cdsousa** [comment]
What we need is a Concrete Syntax Tree (CST), preferably compatible with the current AST.
See, for example, the CST being developed for JavaScript: https://github.com/cst/cst and https://github.com/estree/estree/issues/41.

**Comment by ihnorton** [comment]
x-ref: https://github.com/JuliaLang/julia/issues/21774

**Comment by waldyrious** [comment]
> What we need is a Concrete Syntax Tree (CST), preferably compatible with the current AST.

x-ref: https://github.com/ZacLN/CSTParser.jl
